mobile_verifier/src/speedtests.rs

```diff
@@ -96,580 +87,117 @@ impl SpeedtestDaemon {
-                bytes_per_s(150),
-                70,
-            ),
-            Speedtest::new(
-                parse_dt("2022-08-01 18:00:00 +0000"),
-                bytes_per_s(10),
-                bytes_per_s(118),
-                50,
-            ),
-            Speedtest::new(
-                parse_dt("2022-08-01 12:00:00 +0000"),
-                bytes_per_s(30),
-                bytes_per_s(112),
-                40,
-            ),
-        ];
-        let disjoint_speedtests =
-            speedtests_without_lapsed(disjoint_speedtests.iter(), speedtest_cutoff);
-
-        assert_eq!(contiguous_speedtests.count(), 8);
-        assert_eq!(disjoint_speedtests.count(), 1);
-    }
+pub async fn get_latest_speedtests_for_pubkey<'a>(
+    pubkey: &PublicKeyBinary,
+    exec: &mut Transaction<'_, Postgres>,
+) -> Result<Vec<Speedtest>, sqlx::Error> {
+    let mut speedtests = Vec::new();
+
+    let mut rows = sqlx::query_as::<_, Speedtest>(
+        "SELECT * FROM speedtests where pubkey = $1 order by timestamp desc limit $2",
+    )
+    .bind(pubkey)
+    .bind(SPEEDTEST_AVG_MAX_DATA_POINTS as i64)
+    .fetch(exec);
+
+    while let Some(speedtest) = rows.try_next().await? {
+        speedtests.push(speedtest);
+    }
+    Ok(speedtests)
+}
+
+pub async fn aggregate_epoch_speedtests<'a>(
+    epoch_end: DateTime<Utc>,
+    exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
+) -> Result<EpochSpeedTests, sqlx::Error> {
+    let mut speedtests = EpochSpeedTests::new();
+    // pull the last N most recent speedtests from prior to the epoch end for each pubkey
+    let mut rows = sqlx::query_as::<_, Speedtest>(
+        "select * from (
+            SELECT distinct(pubkey), upload_speed, download_speed, latency, timestamp, row_number()
+            over (partition by pubkey order by timestamp desc) as count FROM speedtests where timestamp < $1
+        ) as tmp
+        where count < $2"
+    )
+    .bind(epoch_end)
+    .bind(SPEEDTEST_AVG_MAX_DATA_POINTS as i64)
+    .fetch(exec);
+    // collate the returned speedtests based on pubkey
+    while let Some(speedtest) = rows.try_next().await? {
+        speedtests
+            .entry(speedtest.pubkey.clone())
+            .or_default()
+            .push(speedtest);
+    }
+    Ok(speedtests)
+}
+
+pub async fn save_speedtest_to_db(
+    report: CellSpeedtestIngestReport,
+    exec: &mut Transaction<'_, Postgres>,
+) -> Result<(), sqlx::Error> {
+    sqlx::query(
+        r#"
+        insert into speedtests (pubkey, upload_speed, download_speed, latency, timestamp)
+        values ($1, $2, $3, $4, $5)
+        "#,
+    )
+    .bind(report.report.pubkey)
+    .bind(report.report.upload_speed as i64)
+    .bind(report.report.download_speed as i64)
+    .bind(report.report.latency as i64)
+    .bind(report.report.timestamp)
+    .execute(exec)
+    .await?;
+    Ok(())
+}
+
+// Clear the speedtests table of tests older than hours defined by SPEEDTEST_LAPSE
+pub async fn clear_speedtests(
```

Comment:
This function should have the same signature as `clear_heartbeats`; you should clear out the speedtests that are before the reward period

Reply:
addressed here: a9b971c45caf62acb2c7b3c439ab824bad3ba01e